File: node/src/scram.ts
import crypto from "node:crypto";

export class ScramExchange {
  private clientNonce: string;
  private clientFirstBare = "";
  private serverSignature?: Buffer;

  constructor(private username: string, nonce?: string) {
    this.clientNonce = nonce ?? generateNonce();
  }

  clientFirstMessage(): string {
    this.clientFirstBare = `n=${escapeValue(this.username)},r=${this.clientNonce}`;
    return `n,,${this.clientFirstBare}`;
  }

  handleServerFirst(password: string, serverFirst: string): string {
    const attrs = parseAttributes(serverFirst);
    const nonce = attrs.r;
    const saltB64 = attrs.s;
    const iterStr = attrs.i;

    if (!nonce || !nonce.startsWith(this.clientNonce)) {
      throw new Error("SCRAM server nonce mismatch");
    }
    if (!saltB64 || !iterStr) {
      throw new Error("SCRAM server-first missing fields");
    }

    const iterations = Number(iterStr);
    const salt = Buffer.from(saltB64, "base64");
    const saltedPassword = hi(password, salt, iterations);
    const clientKey = hmac(saltedPassword, "Client Key");
    const storedKey = sha256(clientKey);

    const clientFinalWithoutProof = `c=biws,r=${nonce}`;
    const authMessage = `${this.clientFirstBare},${serverFirst},${clientFinalWithoutProof}`;

    const clientSignature = hmac(storedKey, authMessage);
    const clientProof = xor(clientKey, clientSignature);
    const serverKey = hmac(saltedPassword, "Server Key");
    this.serverSignature = hmac(serverKey, authMessage);

    return `${clientFinalWithoutProof},p=${clientProof.toString("base64")}`;
  }

  verifyServerFinal(serverFinal: string): void {
    const attrs = parseAttributes(serverFinal);
    const verifier = attrs.v;
    if (!verifier || !this.serverSignature) {
      throw new Error("SCRAM server-final missing verifier");
    }
    const expected = this.serverSignature.toString("base64");
    if (verifier !== expected) {
      throw new Error("SCRAM server signature mismatch");
    }
  }
}

function generateNonce(): string {
  return crypto.randomBytes(18).toString("base64");
}

function escapeValue(value: string): string {
  return value.replace(/=/g, "=3D").replace(/,/g, "=2C");
}

function parseAttributes(message: string): Record<string, string> {
  const attrs: Record<string, string> = {};
  if (!message) return attrs;
  for (const part of message.split(",")) {
    const idx = part.indexOf("=");
    if (idx > 0) {
      attrs[part.slice(0, idx)] = part.slice(idx + 1);
    }
  }
  return attrs;
}

function hi(password: string, salt: Buffer, iterations: number): Buffer {
  return crypto.pbkdf2Sync(password, salt, iterations, 32, "sha256");
}

function hmac(key: Buffer, data: string): Buffer {
  return crypto.createHmac("sha256", key).update(data).digest();
}

function sha256(data: Buffer): Buffer {
  return crypto.createHash("sha256").update(data).digest();
}

function xor(left: Buffer, right: Buffer): Buffer {
  const out = Buffer.alloc(left.length);
  for (let i = 0; i < left.length; i++) {
    out[i] = left[i] ^ right[i];
  }
  return out;
}

File: r/R/scram.R
# ScratchBird-driver
# Copyright (c) 2025-2026 Dalton Calford
#
# Licensed under the Initial Developer's Public License Version 1.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at:
# https://www.firebirdsql.org/en/initial-developer-s-public-license-version-1-0/
sb_scram_client <- function(username) {
  nonce <- openssl::base64_encode(openssl::rand_bytes(18))
  list(
    username = username,
    client_nonce = nonce,
    client_first_bare = "",
    server_signature = NULL
  )
}

sb_scram_client_first <- function(state) {
  escaped <- gsub("=", "=3D", gsub(",", "=2C", state$username, fixed = TRUE), fixed = TRUE)
  state$client_first_bare <- paste0("n=", escaped, ",r=", state$client_nonce)
  list(state = state, message = paste0("n,,", state$client_first_bare))
}

sb_scram_handle_server_first <- function(state, password, server_first) {
  attrs <- parse_scram_attrs(server_first)
  nonce <- attrs$r
  if (is.null(nonce) || !startsWith(nonce, state$client_nonce)) stop("SCRAM server nonce mismatch")
  salt_b64 <- attrs$s
  iter_str <- attrs$i
  if (is.null(salt_b64) || is.null(iter_str)) stop("SCRAM server-first missing fields")
  iterations <- as.integer(iter_str)
  salt <- openssl::base64_decode(salt_b64)
  if (!exists("pbkdf2", where = asNamespace("openssl"))) stop("openssl::pbkdf2 is required for SCRAM")
  salted <- openssl::pbkdf2(password, salt, iterations, keylen = 32, algo = "sha256")
  client_key <- openssl::hmac(salted, "Client Key", algo = "sha256")
  stored_key <- openssl::sha256(client_key)
  client_final_without_proof <- paste0("c=biws,r=", nonce)
  auth_message <- paste(state$client_first_bare, server_first, client_final_without_proof, sep = ",")
  client_signature <- openssl::hmac(stored_key, auth_message, algo = "sha256")
  client_proof <- bitwXor(as.integer(client_key), as.integer(client_signature))
  server_key <- openssl::hmac(salted, "Server Key", algo = "sha256")
  state$server_signature <- openssl::hmac(server_key, auth_message, algo = "sha256")
  proof_b64 <- openssl::base64_encode(as.raw(client_proof))
  list(state = state, message = paste0(client_final_without_proof, ",p=", proof_b64))
}

sb_scram_verify_server_final <- function(state, server_final) {
  attrs <- parse_scram_attrs(server_final)
  verifier <- attrs$v
  if (is.null(verifier)) stop("SCRAM server-final missing verifier")
  expected <- openssl::base64_encode(state$server_signature)
  if (verifier != expected) stop("SCRAM server signature mismatch")
  TRUE
}

parse_scram_attrs <- function(message) {
  attrs <- list()
  parts <- strsplit(message, ",", fixed = TRUE)[[1]]
  for (part in parts) {
    if (part == "") next
    kv <- strsplit(part, "=", fixed = TRUE)[[1]]
    if (length(kv) < 2) next
    attrs[[kv[1]]] <- kv[2]
  }
  attrs
}
